src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -45,15 +101,212 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+// This function does the real work of the thread pool.
+// owned_job is used to tell the differece between the master thread
+// and the worker threads. If owned_job is non-null then it means that
+// we are in the master thread, i.e. the thread that put owned_job in
+// the work queue. This thread, like other worker threads in this function,
+// tries to acquire a lock on the work queue. Each thread then looks for
+// a job to do. In the case of the master thread, it is "owned_job" while
+// a worker thread has to iterate backwards over the work queue and find
+// a job to work on. Once a job has been found, the thread, after some
+// book-keeping, releases the lock and call halide_do_task to do the job.
+// After the job is done, the lock on the queue is re-acquired to update
+// the status and the thread loops again to look for new work. If a worker
+// thread doesn't find any new work, it goes to sleep until awoken by
+// the master thread.
+void goto_work(work *owned_job) {
+    int myjob;
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: goto_work: Trying to get a lock on the work queue\n", tid);
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    work_queue.lock();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+
+        work *job = NULL;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                work_queue.unlock();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                work_queue.lock();
+                continue;
+            } else {
+                myjob = job->next;
```

Comment:
I'd still prefer these to be removed and declare/set myjob = job->next++ below. The reason is that variables that have a long lifetime and get mutated are harder to reason about. If this just gets declared and assigned once, it's simpler to see what the code does.